File: 1_ground_station_ui/src/WifiManager.cpp
#include "WifiManager.h"
#include <QProcess>
#include <QVariantMap>
#include <QSet>
#include <algorithm>

static const char *WIFI_IFACE = "mlan0";
static const char *WPA_CLI = "/usr/sbin/wpa_cli";

QString WifiManager::wpa(const QStringList &args, int timeoutMs) {
    QProcess proc;
    QStringList full;
    full << "-i" << WIFI_IFACE << args;
    proc.start(WPA_CLI, full);
    if (!proc.waitForStarted(timeoutMs))
        return QStringLiteral("ERR: start");
    if (!proc.waitForFinished(timeoutMs)) {
        proc.kill();
        return QStringLiteral("ERR: timeout");
    }
    return QString::fromUtf8(proc.readAllStandardOutput());
}

WifiManager::WifiManager(QObject *parent) : QObject(parent) {
    refreshSaved();
    refreshStatus();
    m_pollTimer.setInterval(3000);
    connect(&m_pollTimer, &QTimer::timeout, this, &WifiManager::refreshStatus);
    m_pollTimer.start();
}

void WifiManager::setBusy(bool v) {
    if (v != m_busy) {
        m_busy = v;
        emit busyChanged();
    }
}

void WifiManager::refreshSaved() {
    QString out = wpa({"list_networks"});
    QVariantList ss;
    const QStringList lines = out.split('\n');
    for (int i = 1; i < lines.size(); ++i) {
        const QStringList parts = lines[i].split('\t');
        if (parts.size() >= 2 && !parts[1].isEmpty())
            ss.append(parts[1]);
    }
    if (ss != m_savedSsids) {
        m_savedSsids = ss;
        emit savedChanged();
    }
}

void WifiManager::refreshStatus() {
    QString out = wpa({"status"});
    QString ssid, state;
    for (const QString &line : out.split('\n')) {
        if (line.startsWith("ssid="))
            ssid = line.mid(5);
        else if (line.startsWith("wpa_state="))
            state = line.mid(10);
    }
    if (ssid != m_currentSsid || state != m_wpaState) {
        m_currentSsid = ssid;
        m_wpaState = state;
        emit statusChanged();
    }
}

void WifiManager::scan() {
    setBusy(true);
    wpa({"scan"});
    QTimer::singleShot(2500, this, &WifiManager::collectScan);
}

void WifiManager::collectScan() {
    QString out = wpa({"scan_results"});
    QVariantList nets;
    QSet<QString> seen;
    const QStringList lines = out.split('\n');
    for (int i = 1; i < lines.size(); ++i) {
        const QStringList parts = lines[i].split('\t');
        if (parts.size() < 5)
            continue;
        const QString bssid = parts[0], freq = parts[1], sig = parts[2],
                      flags = parts[3], ssid = parts[4];
        if (ssid.isEmpty() || seen.contains(ssid))
            continue;
        seen.insert(ssid);
        bool secured = flags.contains("PSK") || flags.contains("WEP") || flags.contains("EAP");
        bool ok = false;
        int sigInt = sig.toInt(&ok);
        if (!ok) sigInt = -100;
        nets.append(QVariantMap{
            {"ssid", ssid}, {"bssid", bssid}, {"freq", freq},
            {"signal", sigInt}, {"secured", secured},
        });
    }
    std::sort(nets.begin(), nets.end(), [](const QVariant &a, const QVariant &b) {
        return a.toMap()["signal"].toInt() > b.toMap()["signal"].toInt();
    });
    m_networks = nets;
    emit networksChanged();
    setBusy(false);
}

void WifiManager::connectTo(const QString &ssid, const QString &password) {
    setBusy(true);
    QString out = wpa({"list_networks"});
    QString netId;
    const QStringList lines = out.split('\n');
    for (int i = 1; i < lines.size(); ++i) {
        const QStringList parts = lines[i].split('\t');
        if (parts.size() >= 2 && parts[1] == ssid) {
            netId = parts[0];
            break;
        }
    }
    if (netId.isEmpty()) {
        QString r = wpa({"add_network"}).trimmed();
        bool ok = false; r.toInt(&ok);
        if (!ok) { setBusy(false); return; }
        netId = r;
        wpa({"set_network", netId, "ssid", QString("\"%1\"").arg(ssid)});
        if (!password.isEmpty())
            wpa({"set_network", netId, "psk", QString("\"%1\"").arg(password)});
        else
            wpa({"set_network", netId, "key_mgmt", "NONE"});
    } else if (!password.isEmpty()) {
        wpa({"set_network", netId, "psk", QString("\"%1\"").arg(password)});
    }
    wpa({"enable_network", netId});
    wpa({"select_network", netId});
    wpa({"save_config"});
    QTimer::singleShot(500, this, &WifiManager::refreshSaved);
    QTimer::singleShot(4000, this, &WifiManager::postConnect);
}

void WifiManager::postConnect() {
    QProcess::startDetached("/sbin/udhcpc",
                            {"-i", WIFI_IFACE, "-q", "-n", "-t", "5"});
    QTimer::singleShot(2000, this, &WifiManager::refreshStatus);
    QTimer::singleShot(2500, this, [this] { setBusy(false); });
}

void WifiManager::disconnect_(const QString &ssid) {
    QString out = wpa({"list_networks"});
    const QStringList lines = out.split('\n');
    for (int i = 1; i < lines.size(); ++i) {
        const QStringList parts = lines[i].split('\t');
        if (parts.size() >= 2 && parts[1] == ssid) {
            wpa({"disable_network", parts[0]});
            break;
        }
    }
    wpa({"disconnect"});
    refreshStatus();
}

File: 1_ground_station_ui/wifi_manager.py
"""WiFiManager: subprocess wpa_cli wrapper exposed to QML.
Inject into skyvision_ui.py at module load time.
"""
import subprocess
import time
import re
from PySide6.QtCore import QObject, Property, Signal, Slot, QTimer

WIFI_IFACE = "mlan0"
WPA_CLI = "/usr/sbin/wpa_cli"


def _wpa(args, timeout=4):
    """Run wpa_cli -i mlan0 <args>, return stdout string."""
    cmd = [WPA_CLI, "-i", WIFI_IFACE] + (args if isinstance(args, list) else args.split())
    try:
        r = subprocess.run(cmd, capture_output=True, text=True, timeout=timeout)
        return r.stdout
    except Exception as e:
        return f"ERR: {e}"


class WiFiManager(QObject):
    networksChanged = Signal()
    statusChanged = Signal()
    busyChanged = Signal()
    savedChanged = Signal()

    def __init__(self, parent=None):
        super().__init__(parent)
        self._networks = []
        self._saved_ssids = []
        self._current_ssid = ""
        self._wpa_state = ""
        self._busy = False
        self._refresh_saved()
        self._refresh_status()
        # poll status every 3s
        self._t = QTimer(self)
        self._t.timeout.connect(self._refresh_status)
        self._t.start(3000)

    # ---- properties ----
    @Property("QVariantList", notify=networksChanged)
    def networks(self):
        return self._networks

    @Property("QVariantList", notify=savedChanged)
    def savedSSIDs(self):
        return self._saved_ssids

    @Property(str, notify=statusChanged)
    def currentSSID(self):
        return self._current_ssid

    @Property(str, notify=statusChanged)
    def wpaState(self):
        return self._wpa_state

    @Property(bool, notify=busyChanged)
    def busy(self):
        return self._busy

    def _set_busy(self, v):
        if v != self._busy:
            self._busy = v
            self.busyChanged.emit()

    # ---- internals ----
    def _refresh_saved(self):
        out = _wpa("list_networks")
        ss = []
        for line in out.splitlines()[1:]:
            parts = line.split("\t")
            if len(parts) >= 2 and parts[1]:
                ss.append(parts[1])
        if ss != self._saved_ssids:
            self._saved_ssids = ss
            self.savedChanged.emit()

    def _refresh_status(self):
        out = _wpa("status")
        ssid = ""
        state = ""
        for line in out.splitlines():
            if line.startswith("ssid="):
                ssid = line.split("=", 1)[1]
            elif line.startswith("wpa_state="):
                state = line.split("=", 1)[1]
        if ssid != self._current_ssid or state != self._wpa_state:
            self._current_ssid = ssid
            self._wpa_state = state
            self.statusChanged.emit()

    # ---- slots invoked from QML ----
    @Slot()
    def scan(self):
        self._set_busy(True)
        _wpa("scan")
        QTimer.singleShot(2500, self._collect_scan)

    def _collect_scan(self):
        out = _wpa("scan_results")
        nets = []
        seen = set()
        for line in out.splitlines()[1:]:
            parts = line.split("\t")
            if len(parts) < 5:
                continue
            bssid, freq, sig, flags, ssid = parts[0], parts[1], parts[2], parts[3], parts[4]
            if not ssid or ssid in seen:
                continue
            seen.add(ssid)
            secured = "PSK" in flags or "WEP" in flags or "EAP" in flags
            try:
                sig_int = int(sig)
            except ValueError:
                sig_int = -100
            nets.append({
                "ssid": ssid,
                "bssid": bssid,
                "freq": freq,
                "signal": sig_int,
                "secured": secured,
            })
        nets.sort(key=lambda n: -n["signal"])
        self._networks = nets
        self.networksChanged.emit()
        self._set_busy(False)

    @Slot(str, str)
    def connectTo(self, ssid, password):
        self._set_busy(True)
        # check if network already exists
        out = _wpa("list_networks")
        net_id = None
        for line in out.splitlines()[1:]:
            parts = line.split("\t")
            if len(parts) >= 2 and parts[1] == ssid:
                net_id = parts[0]
                break
        if net_id is None:
            r = _wpa("add_network").strip()
            if not r.isdigit():
                self._set_busy(False)
                return
            net_id = r
            _wpa(["set_network", net_id, "ssid", f'"{ssid}"'])
            if password:
                _wpa(["set_network", net_id, "psk", f'"{password}"'])
            else:
                _wpa(["set_network", net_id, "key_mgmt", "NONE"])
        else:
            if password:
                _wpa(["set_network", net_id, "psk", f'"{password}"'])
        _wpa(["enable_network", net_id])
        _wpa(["select_network", net_id])
        _wpa("save_config")
        QTimer.singleShot(500, self._refresh_saved)
        # trigger DHCP after association
        QTimer.singleShot(4000, self._post_connect)

    def _post_connect(self):
        subprocess.Popen(["/sbin/udhcpc", "-i", WIFI_IFACE, "-q", "-n", "-t", "5"])
        QTimer.singleShot(2000, self._refresh_status)
        QTimer.singleShot(2500, lambda: self._set_busy(False))

    @Slot(str)
    def disconnect_(self, ssid):
        # disable + select removed
        out = _wpa("list_networks")
        for line in out.splitlines()[1:]:
            parts = line.split("\t")
            if len(parts) >= 2 and parts[1] == ssid:
                _wpa(["disable_network", parts[0]])
                break
        _wpa("disconnect")
        self._refresh_status()
